[AI-6101] Add support for customizable cache keys for persistent cache

AAraKKe opened the pull request:
### What does this PR do?
<!-- A brief description of the change being made with this pull request. -->
<details> <summary>Original implementation </summary>

**The implementation outlined here is the original idea when the PR was opened. After review, the implementation has changed quite a bit.**

This PR adds support for customizable cache keys to be used with the agent persistent cache. This is implemented in the following way:

- A `CacheKey` abstract class that can be extended to implement any cache invalidation logic needed.
- The `AgentCheck` has been modified to rely on a `CacheKey` implementation to provide the key for the persistent cache internally. Using `check_id` is now the default behavior through the `FullConfigCacheKey` implementation but can be overriden when necessary.
- The `logs_persistent_cache_key` can be overriden by a developer that wants to modify what is the `CacheKey` used to handle the logs cursor.
- We are still maintaining some level of separation between checks by adding the check name to the key. It is not the full check_id but, if the developer creates a `CacheKey` that is to simple, the impact of key clashing is reduced to the same kind of check.
- Added 2 implementations of the `CacheKey` interface that work as an example and to provide funcionality:
  - The `FullConfigCacheKey` is equivalent to the previous behavior. It relies on the `check_id` that includes a digest of the full configuration of the check. This default behavior makes this change backward compatible with the usage of `send_log` and `get_log_cursor`, both of them using persistent cache.
  - The `ConfigSetCacheKey` implementation invalidates the cache only when a subset of configuration options have been updated. This provides a similar level of separation between different checks but allowing some configuration options to not invalidate the cache.
- The `CacheKeyManager` allows multiple cache keys to coexist in the `AgentCheck` for differnet purposes. At the moment we only use the `CacheKey` internally to manage the log cursor. However, if we want to include something else we can register a new type of `CacheKeyType` enum and maintain the new one in the manager without having to modify how the `AgentCheck` behaves.

**Note**: the manager forces to have each cache key represented by an enum to avoid relying on simple string keys per cache key. While the intention is to allow to personalize the cache invalidation mechanism, we do not want potentially arbitrary cache keys being register which could trigger cache misses if the string key is mistakenly generated.

</details>

This PR adds support for customizing the persistent cache prefix used when caching any data in the `AgentCheck`. This is done by allowing the developer to override the `persistent_cache_id` method. This method returns the ID that identifies the specific check instance in the Agent persistent cache. 

When a given check returns the same persistent cache ID, the cache keys generated will be stable. By default, the persistent cache ID is the digest of the full check configuration. Any update in the configuration will generate a different ID, causing the cache to be invalidated. This is the current behavior before introducing this feature, which ensures backward compatibility.


### Motivation
Until now, integrations that wanted to emit logs and be consistent between agent restarts would have logs missing or duplicated if the user changed their configuration. This would invalidate the cache and there was limited control over the last cursor. 

The intention of this change is to provide a backward compatible solution that allows developers to modify the cache invalidation behavior in an easy way. 

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Add the `qa/skip-qa` label if the PR doesn't need to be tested during QA.
- [ ] If you need to backport this PR to another branch, you can add the `backport/<branch-name>` label to the PR and it will automatically open a backport PR once this one is merged

ofek commented on datadog_checks_base/datadog_checks/base/utils/cache_key/config_set.py:
```diff
@@ -0,0 +1,54 @@
+        if self.__key is not None:
+            return self.__key
+
+        merged_config = self.check.init_config | self.check.instance
```
1. The agent determines the check ID from only the instance AFAIK (double check) so it would be odd to use the full resolved config. More importantly, there are options that integrations use with the same name that do not have the exact same behavior and therefore this would not be correct.
2. As I mentioned in another comment, the `name` option from the instance config should always be included.

ofek commented on datadog_checks_base/datadog_checks/base/utils/cache_key/config_set.py:
```diff
@@ -0,0 +1,54 @@
+            return self.__key
+
+        merged_config = self.check.init_config | self.check.instance
+        selected_values = tuple(values for key, values in merged_config.items() if key in self.config_options)
+        self.__key = str(hash(self.__sorted_values(selected_values)))
+        return self.__key
+
+    def __sorted_values(self, values: Iterable[Any]) -> tuple[str, ...]:
+        sorted_values = []
+
+        for value in values:
+            if isinstance(value, (list, tuple, set, frozenset)):
+                sorted_values.extend(self.__sorted_values(value))
+            elif isinstance(value, dict):
+                for key, dict_value in value.items():
+                    sorted_values.append(f"{key}:{self.__sorted_values(dict_value)}")
+            else:
+                sorted_values.append(str(value))
+        return tuple(sorted(sorted_values))
```
I think you can get rid of all of this by using the existing `hash_mutable` [utility](https://github.com/DataDog/integrations-core/blob/e4ca80779d4737d36dae52c28c60ca3d05d4e36d/datadog_checks_base/datadog_checks/base/utils/containers.py#L47-L65).

ofek commented on datadog_checks_base/datadog_checks/base/utils/cache_key/manager.py:
```diff
@@ -0,0 +1,76 @@
+        Manager of cache keys for the persistent cache.
+
+        This class defined the different kinds of persistent cache keys to be used when adding and retrieving
+        from the agent persistent cache. The AentCheck can use this manager to to ensure that the correct cache key is
```
```suggestion
        from the agent persistent cache. The AgentCheck can use this manager to to ensure that the correct cache key is
```

AAraKKe commented:
@ofek , answering to your suggestion. I understand that your point is to remove the the need of a class, the rest is essentially the same except for when the initialization happens (maybe?).

I just realized we have a way of running delayed initialization in the check through the `check_initializations` deque. I added the initialization there to avoid having to check for initialization every time we submit a log. The only difference is that the entry point for that is the `run` method. This means that the check needs to be launched for the cache to be initialized. I have checked in all the integrations we have and none of them writes or read from the cache before the run method. I also don't think there is a way for this to happen since the entry point of the check is the `run` method.

About not having a class, I disagree here. The intention of having a class is to allow reusable cache invalidation mechanisms. Partners often have the same way they want to invalidate log cursors. Having a way of encapsulating that logic in a reusable isolated way would be a great benefit for them. The option of not having a class means that the check itself would be littered with logic for cache invalidation logic. It would force developers to create their own custom version of the check if they want to reuse it when the only specialization they want is the logic for cache invalidation. I really think that encapsulating the logic into a class is a better choice here.

The overhead we are talking about here is:
- Create the instance of the class, once per check initialization.
- Call the method in the class to create the string vs just do a string concatenation.

It is true that there is an overhead, but string concatenation is orders of magnitude below the typical latency of check operations to get the logs/metrics which is mostly IO bound.

In any case, even though the overhead here when compared with the operations a check normally does is pretty small I appreciate it still exists. I have reshaped the implementation to have the best of both worlds:

- Rename the approach from a solution to derive cache keys, to a solution to design cache invalidation strategies.
- The instance of the class that encapsulates the logic is created only once, on initialization.
- The invalidation strategy has a `invalidation_token` method that can be overridden to define when the cache will be invalidated.
- The `key_prefix` method generates the key prefix based on the invalidation strategy.
- The `key_prefix` is what is cached on initialization and we avoid the overhead of calling a method on the class to generate the cache key every time.

I think this addresses your concern about performance while at the same time allowing encapsulating different cache invalidation strategies on their own context.

Let me know what you think.

**Note**: I extracted the `check_initializations` calls to its own method which cleans up the `run` method a bit and allows calling it from methods that require the cache to be initialized.

AAraKKe commented on datadog_checks_base/datadog_checks/base/utils/cache_key/full_config.py:
```diff
@@ -0,0 +1,21 @@
+        return self.check.check_id
+
+    def base_key(self) -> str:
+        return self.check.check_id
```
I have modified these classes so the renaming here does not apply anymore I believe. Let me know what do you think.

ofek commented on datadog_checks_base/datadog_checks/base/checks/base.py:
```diff
@@ -1091,7 +1094,12 @@ def entrypoint(self, *args, **kwargs):
 
     def __initialize_persistent_cache_key_prefix(self):
-        self.__persistent_cache_key_prefix = self.cache_invalidation_strategy().key_prefix()
+        if namespace := self.check_id.split(':')[:-1]:
+            namespace = ':'.join(namespace) + ':'
+        else:
+            namespace = ''
```
I think you can just copy the line from my suggestion for simplicity and preventing the variable from being two different types in each assignment https://github.com/DataDog/integrations-core/pull/21316#pullrequestreview-3218113117

```python
    # Remove the instance hash from the check ID
    namespace = ':'.join(self.check_id.split(':')[:-1])
```

<img width="487" height="312" alt="image" src="https://github.com/user-attachments/assets/79bd2eee-867f-45fa-8da3-df6d6df22afb" />

AAraKKe replied:
After discussing online, we do not want to cover the case where the format is not that way. This will never change and is part of the agent internal. Developers should not modify this `check_id`.

ofek reviewed (approved):
Looks good to me now, just final suggestion. Thanks again!
